### ISSUE TITLE
Negative Loss, Transfer Learning/Fine-Tuning Question

### ISSUE THREAD

**Issue by rsomani95** [lucidrains/byol-pytorch#6]
Hi! Thanks for sharing this repo -- really clean and easy to use.

When training using the PyTorch Lightning script from the repo, my loss is negative (and gets more negative over time) when training. Is this expected?
<img width="736" alt="Screenshot 2020-06-22 at 6 23 47 PM" src="https://user-images.githubusercontent.com/17111474/85289735-84e67f00-b4b5-11ea-9459-eddb0a8e494a.png">

---

I'm curious to know if you've fine-tuned a pretrained model using this BYOL as the README example suggested. If yes, how were the results? Any intuition regarding how many epochs to fine-tune for? 

Thanks!

**Comment by nihal94** [comment]
Check this function: [https://github.com/lucidrains/byol-pytorch/blob/e25b5881eee2240a49bf1da2929df8b0ba352fbc/byol_pytorch/byol_pytorch.py#L36] (https://github.com/lucidrains/byol-pytorch/blob/e25b5881eee2240a49bf1da2929df8b0ba352fbc/byol_pytorch/byol_pytorch.py#L36)

I believe it is not the same as in the paper : [https://arxiv.org/pdf/2006.07733.pdf](https://arxiv.org/pdf/2006.07733.pdf)

Check page 28, section G.3 Loss function. Though it is JAX (and not very clear to me), I think the loss function implementations in various repos are not right. But hey, I could be wrong and if someone could shed some light on this I would really appreciate it!

Thanks.

**Comment by lucidrains** [comment]
@nihal94 it should be the same, the loss function is nothing more than the negative of the cosine similarity (negative because we are trying to maximize the similarity)

```python
import torch.nn.functional as F
import torch
from torch import nn

def loss_fn(x, y): 
     x = F.normalize(x, dim=-1, p=2) 
     y = F.normalize(y, dim=-1, p=2) 
     return -2 * (x * y).sum(dim=-1)

def loss_fn_2(x, y):
    return -2 * ((x * y).sum(dim=-1) / (x.norm(dim=-1) * y.norm(dim=-1)))

x = torch.randn(2, 4)
y = torch.randn(2, 4)

l1 = loss_fn(x, y)
l2 = loss_fn_2(x, y)
l3 = -2 * nn.CosineSimilarity(dim=-1)(x, y)

print(l1, l2, l3)
```

**Comment by lucidrains** [comment]
@rsomani95 I haven't yet, I was hoping someone will! Currently doing some transformers work, but will get back to this later this week!

**Comment by nihal94** [comment]
@lucidrains thanks for clarifying. Now it makes sense. Also check the screenshot below, 4th epoch and that's the loss currently

![image](https://user-images.githubusercontent.com/13083807/85314841-afe1ca80-b4d7-11ea-8246-198256718c57.png)

**Comment by lucidrains** [comment]
@nihal94 yea, maybe I should normalize the loss so it lies between 0 and -1, just so it doesn't scare people into thinking it is broken

**Comment by lucidrains** [comment]
@tiredrandomuser it shouldn't make a difference to training anyhow

**Comment by lucidrains** [comment]
ok done! https://github.com/lucidrains/byol-pytorch/commit/1cce49dc21399c0fef0898be366556aaea1739c3 losses should fall around [-4, 0] now

**Comment by tiredrandomuser** [comment]
@lucidrains yes. I'm yet to check the model. Will update here soon if possible. And thanks for the quick and clean implementation

**Comment by rsomani95** [comment]
> @rsomani95 I haven't yet, I was hoping someone will! Currently doing some transformers work, but will get back to this later this week!

Awesome. I should have some more time on my hands in a couple of weeks to play around with this too. The biggest hurdle is compute time. A dataset of 120,000 images takes about 1.5 hrs per epoch on a  K40 on colab.

**Comment by lucidrains** [comment]
@rsomani95 let us know! finally it seems self-supervised learning could be accessible to us mere mortals lol

**Comment by rsomani95** [comment]
Hahahaha that's precisely how I feel

**Comment by lucidrains** [comment]
solved with https://github.com/lucidrains/byol-pytorch/pull/7 !

**Comment by AderonHuang** [comment]
good
